public/src/modules/translator.js

```diff
@@ -311,7 +311,28 @@
 
 			if (key) {
 				return translation.then(function (x) {
-					return x[key];
+					if (typeof x[key] === 'string') return x[key];
+					const keyParts = key.split('.');
+					for (let i = 0; i <= keyParts.length; i++) {
+						if (i === keyParts.length) {
+							// default to trying to find key with the same name as parent or equal to empty string
+							return x[keyParts[i - 1]] !== undefined ? x[keyParts[i - 1]] : x[''];
+						}
+						switch (typeof x[keyParts[i]]) {
+							case 'object':
+								x = x[keyParts[i]];
+								break;
+							case 'string':
+								if (i === keyParts.length - 1) {
+									return x[keyParts[i]];
+								}
+
+								return false;
```

Comment:
Shouldn't this be continue to support nested objects that are deeper than 2 levels? IE
```
{  
   "key": {
        "key1": { 
             "key2": "some translation" 
         }
   }
}
```

Can you add some tests for the above? I added a sample one to add custom translations here https://github.com/NodeBB/NodeBB/commit/7928036a1b9f9d6891c90842222209539b9ebfae

Reply:
Specifically add a test like the below

```
it('should handle nested translations', async function () {
			shim.addTranslation('en-GB', 'my-namespace', {
				key: {
					key1: 'key1 translated',
					key2: {
						key3: 'key3 translated',
					},
				},
			});
			const t1 = await shim.translate('this is best [[my-namespace:key.key1]]');
			const t2 = await shim.translate('this is best [[my-namespace:key.key2.key3]]');
			assert.strictEqual(t1, 'key1 translated');
			assert.strictEqual(t2, 'key3 translated');
		});
```

Reply:
Added two nested translation tests - one you suggested and another one to ensure defaults work to allow for partial keys when a key is nested.

Reply:
Ah and about continuing - this return false (which causes the translation to fail, default to the key name and write that no translation was found in logs) will only happen if the translation reaches a string while it still has sub-keys to go through. So in the example you gave - it will return false if you try to translate key.key1.key2.key3 where key2 is already a string so can't be traversed further.
Same thing (return false) happens if the type is neither string nor object.

continue isn't needed, as these returns can only be hit if the translation at current keyPart isn't an object that could be traversed further.